# Exploring eBay Car Sales Data
We'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
You can find the dataset [here](https://data.world/data-society/used-cars-data).

<center>The data dictionary provided with data is as follows:

| <center>Column | <center>Description |
|---:|:---|
| `dateCrawled` | When this ad was first crawled. All field-values are taken from this date |
| `name` | Name of the car |
| `seller` | Whether the seller is private or a dealer |
| `offerType` | The type of listing |
| `price` | The price on the ad to sell the car |
| `abtest` | Whether the listing is included in an A/B test |
| `vehicleType` | The vehicle Type |
| `yearOfRegistration` | The year in which the car was first registered |
| `gearbox` | The transmission type |
| `powerPS` | The power of the car in PS |
| `model` | The car model name |
| `kilometer` | How many kilometers the car has driven |
| `monthOfRegistration` | The month in which the car was first registered |
| `fuelType` | What type of fuel the car uses |
| `brand` | The brand of the car |
| `notRepairedDamage` | If the car has a damage which is not yet repaired |
| `dateCreated` | TThe date on which the eBay listing was created |
| `nrOfPictures` | The number of pictures in the ad |
| `postalCode` | The postal code for the location of the vehicle |
| `lastSeen` | When the crawler saw this ad last online |

### Project Goal
The aim of this project is to clean the data and analyze the used car listings.

### Knowledge used for the project

 - `Pandas.read_csv()` **function** with `encoding='Latin-1'` **parameter** | to open and read a .csv file;
 - `Pandas.Series()` **function** with `encoding='Latin-1'` **parameter** | to create a new Series; 
 - `Pandas.DataFrame()` **function** with `columns=['mean_price']` **parameter** | to create a new DataFrame from a Series specifying the column name; 
 - `DataFrame.info()` **method** | To have a DataFrame overview;
 - `DataFrame.head()` **method** | To print first 5 rows;
 - `DataFrame.columns` **attribute** | To return the label of each column;
 - `DataFrame.describe()` **method** with `include='all'` **parameter** | To understand the distribution without any extra data processing with null values included;
 - `DataFrame.drop()` **method** with `axis=1` **parameter** | To remove unnecessary <span style='color:Blue'> columns  </span> of our dataset;
 - `Series.unique()` **method** | To return all unique values of the column;
 - `Series.shape` **attribute** | To return the number of rows and columns;
 - `Series.apply()` **method** | To suppress usage of scientific notation;
 - `Series.value_counts()` **method** with `normalize=True` & `dropna=False` | To count distinct values in the series - Normalize: If True then the object returned will contain the relative frequencies of the unique values & dropna False to include counts of NaN;
 - `Series.sort_index()` **method** with `ascending=False` **parameter** | To suppress usage of scientific notation;
 - `Series.between()` **method** | To filter series by value range;
 - `Series.str()` **method** | To perform vectorized string functions without using for loop;
 - `Series.size` **attribute** | To filter series by value range;
 - `Series.mean()` **method** | To return series mean;
 - `Series.loc()` **method** | To filter series rows by boolean and columns by column name; 
 - `Series.index` **attribute** | To store into a list the DataFrame index.
 - `Series.map()` **method** | To use a from-to dictionary and overite columns data; 
 - `Series.astype()` **method** | To convert a string to integer type; 
 - `Series.isnull()` **method** | To filter series rows that only contain null values; 
 - `Series.replace()` **method** | To replace string values; 
 - `Series.split()` **method** | To split string values based in a specific content, then selecting the first element. 

### 1. Importing libraries and reading our file
We will use NunPy and Pandas libraries and import the csv file

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('C:/Users/Daniel Mendes/Projects/Databases/autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

### 2. First impressions
We used .info() and .head() methods for a initial look at the database and we noticed:

    1. Many columns with null data: vehicleType, gearbox, model, fuelType and notRepairedDamage. This last one with significant amount of null values;
    2. Upper and lower cases in columns names;
    3. Date columns imported as string (object).
    4. The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

### 3. Standardizing our columns
Let's start our data cleaning changing and standardize our columns.

In [2]:
autos.columns = ['date_crawled',
                 'name',
                 'seller',
                 'offer_type',
                 'price',
                 'ab_test',
                 'vehicle_type',
                 'registration_year',
                 'gear_box',
                 'power_ps',
                 'model',
                 'kilometer',
                 'registration_month',
                 'fuel_type',
                 'brand',
                 'unrepaired_damage',
                 'ad_created',
                 'nr_of_pictures',
                 'postal_code',
                 'last_seen']

print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


With our new columns, lets now investigate now more our data using .describe() method.

In [3]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year gear_box       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

### 4. Removing columns

We can note that 'privat' seller represents 99,99% of the total seller types, something similar happens with offer_type column that have only 2 unique values and one of them represents over 95% of the total. Those columns are strong candidates to be dropped, since they dont offer us any insights. Also all the data in num_pictures is 0, it appears that none of the listings contain any pictures.

Let's drop those three columns and  deepdive in price column to check if our integer column is okay.

In [4]:
autos.drop(['seller','offer_type','nr_of_pictures'], axis=1)

print('Unique values:')
print(autos['price'].unique().shape, '\n')
print('Descriptive statistics:')
print(autos['price'].describe().apply(lambda x: format(x, 'f')), '\n')
print('Top 5 values:')
print(autos['price'].value_counts().sort_index(ascending=False).head())

Unique values:
(5597,) 

Descriptive statistics:
count        371528.000000
mean          17295.141865
std         3587953.744410
min               0.000000
25%            1150.000000
50%            2950.000000
75%            7200.000000
max      2147483647.000000
Name: price, dtype: object 

Top 5 values:
2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
Name: price, dtype: int64


### 5. Translating Non-English Words

In total, there were six columns recorded in German. However, we had already dropped two of the affected columns. We will explore the remaining four below

In [5]:
german_columns = ['vehicle_type', 'gear_box','fuel_type', 'unrepaired_damage']

# print the unique german words in each column of interest
for column in german_columns:
    print('')
    print(autos[column].unique())


[nan 'coupe' 'suv' 'kleinwagen' 'limousine' 'cabrio' 'bus' 'kombi'
 'andere']

['manuell' 'automatik' nan]

['benzin' 'diesel' nan 'lpg' 'andere' 'hybrid' 'cng' 'elektro']

[nan 'ja' 'nein']


In [6]:
# translator function
def translate_column(df,column,value_dict):
    df[column] = df[column].map(value_dict)

# A dictionary of translations
translations ={# vehicle type terms
               'bus':'bus',
               'limousine':'limousine',
               'kleinwagen': 'small car',
               'kombi': 'combo',
               'coupe':'coupe',
               'suv':'suv',
               'cabrio': 'cabrio',
               'andere': 'other',
               'cabrio': 'convertible',
               # gear box terms
               'manuell': 'manual',
               'automatik': 'automatic',
               # fuel type terms
               'lpg':'lpg',
               'benzin': 'gasoline',
               'diesel':'diesel',
               'cng':'natural gas',
               'hybrid':'hybrid',
               'elektro':'electric',
               'andere':'other',
               # unrepaired damage terms
               'nein':'no',
               'ja':'yes'}

Let's translate each of these columns. We will also print out the value counts before and after translation to ensure that the process ran adequately, without compromising our data:

In [7]:
for item in german_columns:
    print('before translating', '\n')
    print(autos[item].value_counts(dropna=False))
    
    print('\n','after translating')
    translate_column(autos, item, translations)
    print(autos[item].value_counts(dropna=False))  
    print('-'*30)

before translating 

limousine     95894
kleinwagen    80023
kombi         67564
NaN           37869
bus           30201
cabrio        22898
coupe         19015
suv           14707
andere         3357
Name: vehicle_type, dtype: int64

 after translating
limousine      95894
small car      80023
combo          67564
NaN            37869
bus            30201
convertible    22898
coupe          19015
suv            14707
other           3357
Name: vehicle_type, dtype: int64
------------------------------
before translating 

manuell      274214
automatik     77105
NaN           20209
Name: gear_box, dtype: int64

 after translating
manual       274214
automatic     77105
NaN           20209
Name: gear_box, dtype: int64
------------------------------
before translating 

benzin     223857
diesel     107746
NaN         33386
lpg          5378
cng           571
hybrid        278
andere        208
elektro       104
Name: fuel_type, dtype: int64

 after translating
gasoline       223857
diesel

### 6. Identifying pattern in name column

the name column is a long string and can give us some additional information. For example, we can see that in the `model` column there are some empty values that we can filled with the first string before the underscore of `name` column to enhance our DataFrame. So when the `model` column is missing, we fill with this key word from `name`

In [8]:
model_by_name = autos['name'].str.split('_').str[0]
autos['model_by_name'] = model_by_name
autos.loc[autos['model'].isnull(),'model'] = autos['model_by_name']

print(model_by_name)
print(autos[['name','model']].head())

0               Golf
1                 A5
2               Jeep
3               GOLF
4              Skoda
             ...    
371523         Suche
371524         Smart
371525    Volkswagen
371526            VW
371527           BMW
Name: name, Length: 371528, dtype: object
                             name  model
0                      Golf_3_1.6   golf
1            A5_Sportback_2.7_Tdi     A5
2  Jeep_Grand_Cherokee_"Overland"  grand
3              GOLF_4_1_4__3TÜRER   golf
4  Skoda_Fabia_1.4_TDI_PD_Classic  fabia


### 7. Removing 'price' rows outliers

As observed earlier, the `25th`, `50th` and `75th` percentile of the prices are `1,100`, `2,950` and `7,200`. We will use a selection method to create a “data fence” from our 25th to 75th percentiles. Any data beyond this fence will be considered an outlier.

We already have enough information to calculate our interquartile range IQR. We will use the equation below to determine our upper and lower price fences:

    Interquartile Range (IQR) = 75th Percentile - 25th Percentile
    Lower fence = 25th Percentile - 1.5*(IQR)
    Upper fence = 75th Percentile + 1.5*(IQR)
    
Lets define all these boundaries in the code cell below and remove the limits:

In [9]:
price_25, price_75 = (1100, 7200)

IQR = price_75 - price_25
lower_fence = price_25 - (1.5*IQR)
upper_fence = price_75 + (1.5*IQR)

print('Interquartile Range: ', IQR)
print('Lower fence: ', lower_fence)
print('Upper fence: ', upper_fence)

autos = autos[autos['price'].between(lower_fence, upper_fence)]
print(autos['price'].unique().shape)

Interquartile Range:  6100
Lower fence:  -8050.0
Upper fence:  16350.0
(3402,)


### 8. Investigating date columns

As we mentioned in the first impressions section, there are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary.

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like .describe() to understand the distribution without any extra data processing.

In [10]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21
5  2016-04-04 17:36:23  2016-04-04 00:00:00  2016-04-06 19:17:07

We can see that the same pattern is used in the 3 fields, they use full timestamp. Let's just take the dates for now and understand the range.

In [11]:
print('Total unique days:',autos['date_crawled'].str[:10].unique().size)
print('Top 5 dates:')
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head())

Total unique days: 34
Top 5 dates:
2016-03-05    0.025702
2016-03-06    0.014494
2016-03-07    0.035738
2016-03-08    0.033558
2016-03-09    0.034233
Name: date_crawled, dtype: float64


In [12]:
print('Total unique days:',autos['ad_created'].str[:10].unique().size)
print('Top 5 dates:')
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head())

Total unique days: 109
Top 5 dates:
2014-03-10    0.000003
2015-03-20    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
2015-08-10    0.000003
Name: ad_created, dtype: float64


In [13]:
print('Total unique days:',autos['last_seen'].str[:10].unique().size)
print('Top 5 dates:')
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head())

Total unique days: 34
Top 5 dates:
2016-03-05    0.001350
2016-03-06    0.004357
2016-03-07    0.005483
2016-03-08    0.008437
2016-03-09    0.010431
Name: last_seen, dtype: float64


Now let's investigate `registration_year` column, already as number.

In [14]:
autos.registration_year.describe()

count    343588.000000
mean       2004.037661
std          89.287534
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2007.000000
max        9999.000000
Name: registration_year, dtype: float64

We can see that this field clearly have outliers values. Minimum value is the year of 1000, when cars wasen't invented which is 1886. We aso have dates afther the current year, like 9999.
In this case we will need to set maximum and minimum limits, and in this case we can use qualitative reasons:

    * Minimum: 1908. The year when one of the first cars accessible to the masses was developed (Model T), in 1908 by Ford Motor Company.
    * Maximum: 2021. Last year.

In [15]:
autos = autos[autos['registration_year'].between(1908, 2021)]
autos.registration_year.describe()

count    343421.000000
mean       2002.907099
std           7.443518
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2007.000000
max        2019.000000
Name: registration_year, dtype: float64

### 9. Converting data to integer

As we saw, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by Pandas. Lets convert in them to integer using only the 10 first characters so dates become uniform numeric data. "2016-03-21" becomes the integer 20160321.

In [16]:
#Removing '-' character and selecting only the first 8 characters
autos['date_crawled'] = autos['date_crawled'].str.replace('-','').str[:8]
autos['last_seen'] = autos['last_seen'].str.replace('-','').str[:8]
autos['ad_created'] = autos['ad_created'].str.replace('-','').str[:8]

#Converting string to integer
autos['date_crawled'] = autos['date_crawled'].astype(int)
autos['last_seen'] = autos['last_seen'].astype(int)
autos['ad_created'] = autos['ad_created'].astype(int)

print(autos[['date_crawled', 'last_seen', 'ad_created']].head())
print('\n')
print(autos[['date_crawled', 'last_seen', 'ad_created']].info())

   date_crawled  last_seen  ad_created
0      20160324   20160407    20160324
2      20160314   20160405    20160314
3      20160317   20160317    20160317
4      20160331   20160406    20160331
5      20160404   20160406    20160404


<class 'pandas.core.frame.DataFrame'>
Int64Index: 343421 entries, 0 to 371526
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype
---  ------        --------------   -----
 0   date_crawled  343421 non-null  int32
 1   last_seen     343421 non-null  int32
 2   ad_created    343421 non-null  int32
dtypes: int32(3)
memory usage: 6.6 MB
None


### 10. Analysing data by brand

When working with data on cars, it's natural to explore variations across different car brands. We will use aggregation to understand the brand column.

In [17]:
unique_brands = autos['brand'].value_counts()
top_brands = unique_brands[:10].index
print(unique_brands)

volkswagen        74747
opel              39608
bmw               34968
mercedes_benz     30168
audi              27672
ford              24744
renault           17842
peugeot           10909
fiat               9600
seat               6771
mazda              5530
skoda              5312
smart              5237
citroen            5105
nissan             4809
toyota             4523
hyundai            3480
sonstige_autos     3139
volvo              3097
mitsubishi         2982
mini               2907
honda              2787
kia                2360
suzuki             2310
alfa_romeo         2275
chevrolet          1686
chrysler           1418
dacia               895
daihatsu            806
subaru              738
jeep                631
porsche             605
trabant             586
daewoo              542
saab                516
rover               484
land_rover          477
lancia              468
jaguar              462
lada                225
Name: brand, dtype: int64


Lets filter and only analyse the top 10 brands. Lets use for loop to get the mean value for each of the top 10 brands.

In [18]:
brand_price_dictionary = {}
for brand in top_brands:
    mean_price = autos.loc[autos['brand'] == brand,'price'].mean()
    brand_price_dictionary[brand] = round(mean_price,2)
for item in brand_price_dictionary:
    print(item,":",brand_price_dictionary[item])

volkswagen : 3972.72
opel : 2611.43
bmw : 5502.6
mercedes_benz : 5151.43
audi : 5476.94
ford : 2917.89
renault : 2196.4
peugeot : 2995.38
fiat : 2635.25
seat : 3716.63


We observed that in the top 10 brands, there's a distinct price gap.

 - BMW, Audi and Mercedes Benz are more expensive
 - Renault, Opel, Fiat, Ford and Peugeot are less expensive
 - Volkswagen and Seat is in between
 
Lets do the same for kilometers to see if we have significant changes in cars usage between the brands:

In [19]:
brand_km_dictionary = {}
for brand in top_brands:
    mean_km = autos.loc[autos['brand'] == brand,'kilometer'].mean()
    brand_km_dictionary[brand] = round(mean_km)
for item in brand_km_dictionary:
    print(item,":",brand_km_dictionary[item])

volkswagen : 132652
opel : 130003
bmw : 138925
mercedes_benz : 137352
audi : 138574
ford : 126102
renault : 128636
peugeot : 125839
fiat : 117348
seat : 124360


Now that we have both dictionarys, lets create first a panda series and a panda dataframe with our first dictionary. Then convert the other dictionary into a Series to add as a new column in the brand new DataFrame.

In [20]:
brand_price_series = pd.Series(brand_price_dictionary)
new_dataframe = pd.DataFrame(brand_price_series, columns=['mean_price'])
brand_km_series = pd.Series(brand_km_dictionary)
new_dataframe['mean_km'] = brand_km_series
print(new_dataframe)

               mean_price  mean_km
volkswagen        3972.72   132652
opel              2611.43   130003
bmw               5502.60   138925
mercedes_benz     5151.43   137352
audi              5476.94   138574
ford              2917.89   126102
renault           2196.40   128636
peugeot           2995.38   125839
fiat              2635.25   117348
seat              3716.63   124360


### 11. Finding the most common brand/model combinations

In [21]:
common_model_dic = {}

for brand in top_brands:
    common_model = autos.loc[autos['brand'] == brand,'model'].value_counts(ascending=False).index[0]
    common_model_dic[brand] = common_model
    
common_model_series = pd.Series(common_model_dic)
new_dataframe['most_common_model'] = common_model_series
print(new_dataframe)

               mean_price  mean_km most_common_model
volkswagen        3972.72   132652              golf
opel              2611.43   130003             corsa
bmw               5502.60   138925               3er
mercedes_benz     5151.43   137352          c_klasse
audi              5476.94   138574                a4
ford              2917.89   126102             focus
renault           2196.40   128636            twingo
peugeot           2995.38   125839           2_reihe
fiat              2635.25   117348             punto
seat              3716.63   124360             ibiza


### 12. Checking if mileage affects price

First, lets have a overview of our `kilometer` column.

In [22]:
print(autos['kilometer'].describe())
print('\n')
print(autos['kilometer'].unique())

count    343421.000000
mean     129466.675014
std       36878.600018
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer, dtype: float64


[150000 125000  90000  40000  30000   5000 100000  60000  70000  80000
  50000  20000  10000]


Looks like the column is already grouped by some mileage ranges, so we don't need to group the data for this analysis. Now, we will get the mean price value for each mileage range.

In [23]:
mileage_dic = {}
unique_km = autos['kilometer'].unique()

for item in unique_km:
    mean_price = autos.loc[autos['kilometer'] == item,'price'].mean()
    mileage_dic[item] = round(mean_price,2)
    
for item in mileage_dic:
    print('KM',item,":",'$',mileage_dic[item])

KM 150000 : $ 3183.64
KM 125000 : $ 4671.62
KM 90000 : $ 5930.84
KM 40000 : $ 8263.82
KM 30000 : $ 7767.99
KM 5000 : $ 2342.63
KM 100000 : $ 5366.39
KM 60000 : $ 7424.94
KM 70000 : $ 6903.88
KM 80000 : $ 6354.74
KM 50000 : $ 7790.96
KM 20000 : $ 6926.13
KM 10000 : $ 7664.88


Mileage doesn't seem to have exactly a pattern and a very strong correlation with the price of cars. Especially the 5,000km cars that have a considerably lower price than the others. My suspicion looking at this information is that the amount of cars with 5,000km is low and by randomness they can be cheaper cars (worse brands or inferior models). Or the sample is also small and there are cars with incorrect and/or very low values, which can bias and pull the average value down.

Lets confirm or deny by looking by counting the amount of cars for each mileage cohort.

In [24]:
autos['kilometer'].value_counts()

150000    235329
125000     35266
100000     14219
90000      10886
80000       9290
70000       7811
60000       6654
5000        6133
50000       5430
40000       4220
30000       3805
20000       3423
10000        955
Name: kilometer, dtype: int64

Apparently our hypothesis was wrong, there are not so few records with 5,000 km cars. Another hypothesis is that cars from better brands (we have already seen that the brand has a strong correlation with the price) usually stay longer with the owner and are only offered for sale after having driven many more kilometers.

Let's group the brands we've seen that have the most expensive cars into "premium brands" and see how much they represent within each mileage cohort.

In [25]:
premium_brands_mileage_dic = {}

for item in unique_km:
    premium = autos.loc[
          (autos['kilometer'] == item)
          & ((autos['brand'] == 'bmw') | (autos['brand'] == 'audi') | (autos['brand'] == 'mercedes_benz')),:].size
    total = autos[autos['kilometer'] == item].size
    penetration = premium / total * 100
    premium_brands_mileage_dic[item] = round(penetration,2)

for item in premium_brands_mileage_dic:
    print('KM',item,":",premium_brands_mileage_dic[item],'%')

KM 150000 : 31.88 %
KM 125000 : 19.44 %
KM 90000 : 16.31 %
KM 40000 : 9.22 %
KM 30000 : 12.09 %
KM 5000 : 22.21 %
KM 100000 : 17.54 %
KM 60000 : 12.74 %
KM 70000 : 13.52 %
KM 80000 : 15.83 %
KM 50000 : 9.87 %
KM 20000 : 12.77 %
KM 10000 : 9.53 %


Again our hypothesis does not prove to be true, due to the high concentration of luxury brand cars being within the 5,000km group (2nd largest concentration).

Another important variable in the price composition may be the year of registration of the vehicle, the older ones may tend to be cheaper. Let's see if in fact the year has a significant correlation with the price.

Below we will group the years between `before_90s`, `year_90s`, `year_00s` and `after_2010`.

In [26]:
old_cars_dic = {}

before_90s = autos.loc[autos['registration_year'] < 1990,'price'].mean()
year_90s = autos.loc[autos['registration_year'].between(1990,1999),'price'].mean()
year_00s = autos.loc[autos['registration_year'].between(2000,2009),'price'].mean()
after_2010 = autos.loc[autos['registration_year'] >= 2010,'price'].mean()

old_cars_dic['before_90s'] = round(before_90s,2)
old_cars_dic['year_90s'] = round(year_90s,2)
old_cars_dic['year_00s'] = round(year_00s,2)
old_cars_dic['after_2010'] = round(after_2010,2)
    
for item in old_cars_dic:
    print(item,":",'$',old_cars_dic[item])

before_90s : $ 4202.53
year_90s : $ 1474.16
year_00s : $ 4469.5
after_2010 : $ 6608.53


Excellent. It looks like we actually have a strong correlation. Also, vintage cars are cheaper to some extent, as pre-90s cars appreciate in value, likely for their scarcity and collector car profile.

Let's use these same groups to create a DataFrame containing a table of frequency in relation to the percentage of representation of each of these groups. With the year groups on the axis and the mileage on the column.

In [27]:
before_90s_dic = {}
year_90s_dic = {}
year_00s_dic = {}
after_2010_dic = {}

for item in unique_km:
    recent = autos.loc[
          (autos['kilometer'] == item)
          & (autos['registration_year'] < 1990),:].size
    total = autos[autos['kilometer'] == item].size
    penetration = recent / total * 100
    before_90s_dic[item] = round(penetration,2)
    
for item in unique_km:
    recent = autos.loc[
          (autos['kilometer'] == item)
          & (autos['registration_year'].between(1990,1999)),:].size
    total = autos[autos['kilometer'] == item].size
    penetration = recent / total * 100
    year_90s_dic[item] = round(penetration,2)
    
for item in unique_km:
    recent = autos.loc[
          (autos['kilometer'] == item)
          & (autos['registration_year'].between(2000,2009)),:].size
    total = autos[autos['kilometer'] == item].size
    penetration = recent / total * 100
    year_00s_dic[item] = round(penetration,2)
    
for item in unique_km:
    recent = autos.loc[
          (autos['kilometer'] == item)
          & (autos['registration_year'] >= 2010),:].size
    total = autos[autos['kilometer'] == item].size
    penetration = recent / total * 100
    after_2010_dic[item] = round(penetration,2)

In [28]:
before_90s_series = pd.Series(before_90s_dic)
year_by_mile_df = pd.DataFrame(before_90s_series, columns=['before_90s'])

year_90s_series = pd.Series(year_90s_dic)
year_by_mile_df['year_90s'] = year_90s_series

year_00s_series = pd.Series(year_00s_dic)
year_by_mile_df['year_00s'] = year_00s_series

after_2010_series = pd.Series(after_2010_dic)
year_by_mile_df['after_2010'] = after_2010_series

print(year_by_mile_df)

        before_90s  year_90s  year_00s  after_2010
150000        1.90     32.54     55.78        9.77
125000        2.15     18.58     65.28       13.99
90000         3.43     13.87     60.03       22.67
40000         5.07      5.17     26.47       63.29
30000         5.28      9.09     21.73       63.89
5000         13.45     29.46     34.88       22.21
100000        5.09     15.66     61.31       17.94
60000         4.66      7.32     43.58       44.44
70000         4.25      9.38     51.04       35.32
80000         4.17     11.55     56.34       27.94
50000         5.03      6.06     34.83       54.09
20000         5.78     13.53     26.15       54.54
10000        13.19      7.75     16.96       62.09


### Conclusion

We realized that the `brand` is an important variable in the composition of the vehicle's `price`, as well as the `year_of_registration`. For `kilometers`, the correlation with price is not so clear and we found that the unusually low price for cars in the 5,000km range is due to the high concentration of 90s cars in this group.